src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs

```diff
@@ -514,16 +514,34 @@ public static FileAttributes GetAttributes(string fullPath)
             return attributes;
         }
 
+        public static FileAttributes GetAttributes(SafeFileHandle fileHandle)
+        {
+            ThrowHelper.ThrowForMissingPath_SafeFileHandle(fileHandle.Path);
+            return GetAttributes(fileHandle.Path!);
```

Comment:
Now that we know that SafeFileHandle is available in Unix, is there a way to implement the new APIs in a way that they don't just defer the job to the string overloads?

Reply:
I'm not sure about that... 
@danmoseley Do you know something about this?

Reply:
GetAttributes uses [`stat`](https://man7.org/linux/man-pages/man2/stat.2.html) to get the information about the given path. You will need to write a similar method that uses [`fstat`](https://linux.die.net/man/2/fstat) to get the information of the given file handle.
Similarly in SetAttributes, the syscalls used take a path, you need to investiga if there are equivalent versions that do the same with the file handle.

cc @tmds in case he has a better idea.

Reply:
@tmds Can you confirm this?

Reply:
You want to use `fstatat` with `AT_SYMLINK_NOFOLLOW`.
If what you stat is a symbolic link. you want to make an additional call without `AT_SYMLINK_NOFOLLOW` to see if the target is a directory.
For `FileAttributes.ReadOnly` you can refactor `FileStatus.IsModeReadOnlyCore` so it is a `static` method that you can also use.

For setting attributes you want to use `fchmod` for `FileAttributes.ReadOnly`. And for `FileAttributes.Hidden` (on macOS) I think you want to keep using `lchflags` because I assume `fchflags` will affect the target of the link.

Reply:
> You want to use fstatat with AT_SYMLINK_NOFOLLOW.

Actually, this doesn't take the file as a handle. So for getting the attributes you can call into the existing implementation.

For setting you could use `fchmod` instead of `chmod`.

Reply:
Maybe with these APIs it makes sense to follow the links?

If that is so, then you can use `fstat`, `fchflags`, and `fchmod`.

Reply:
> Maybe with these APIs it makes sense to follow the links?

@deeprobin what do these APIs do when the path is a symbolic link on Windows. Do they reflect the link entry, or the target entry?

Reply:
> You want to use `fstatat` with `AT_SYMLINK_NOFOLLOW`. If what you stat is a symbolic link. you want to make an additional call without `AT_SYMLINK_NOFOLLOW` to see if the target is a directory.
> 
> For determining `FileAttributes.ReadOnly` you can refactor `FileStatus.IsModeReadOnlyCore` so it is a `static` method that you can also use.
> 
> For setting attributes you want to use `fchmod` for `FileAttributes.ReadOnly`. And for `FileAttributes.Hidden` (on macOS) I think you want to keep using `lchflags` because I assume `fchflags` will affect the target of the link.

I have never done Unix-specific implementations. Is there anything to consider or how can I use `fstat`, ... use?

Reply:
> > Maybe with these APIs it makes sense to follow the links?
> 
> @deeprobin what do these APIs do when the path is a symbolic link on Windows. Do they reflect the link entry, or the target entry?

@danmoseley Do you know the actual behavior of `SetAttributes(string path, FileAttributes attributes)`?

Reply:
> Do they reflect the link entry, or the target entry?

You can check by doing something like:
```cs
var file = Path.GetTempFileName();
var link = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());

var fileHandle = File.OpenHandle(file, FileMode.Open);
var linkHandle = File.OpenHandle(file, FileMode.Open);

File.SetLastWriteTime(linkHandle, DateTime.Now.AddYears(-1));

Console.WriteLine(File.GetLastWriteTime(linkHandle));
Console.WriteLine(File.GetLastWriteTime(fileHandle));
```

I assume this is controlled by the `FILE_FLAG_OPEN_REPARSE_POINT`?

The implementation is setting this flag for the existing methods that accept `string path`, which means they'd affect the link entry.

`File.OpenHandle` does not set this flag, so the new methods that accept `SafeFileHandle fileHandle` will affect the target entry.

We want the API to behave the same on Linux. So we can use `fstat`, `fchflags`, and `fchmod` because those would affect the target (`open` follows the links, similar to `CreateFile` without `FILE_FLAG_OPEN_REPARSE_POINT`).

Reply:
Symbolic links throw a `FileNotFoundException`.

See [sharplab example](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmABgFgAoUgRmqLIAIjaA6ASQHkBua6gG4BDKEwBmASwA2MJgF4mABSEYAFmwDiMDABUYAWwAOAMWkwAckP0wAFAEo+VYaKkS8Aa3lKV6gMIR9YDdbZTVNbT0jUNV7NG8wrQwAJSE8ABMA0xlLa3sHfioJMRsAQiyYNgBRBAkAZwxam0kZOzsmAG9qJm6mcrZfWBVbZph8qgBfAqLSvuq6hptXD1aOrp6+gZghgGUAT0CIVzAAGTd3RbO4kbHJqmo+7e1joXqAdSgJDBgdCVyl9ziABEhj9rGxzBAAO5sACCaTSAE0tlBGnBaK1HAxaABOGx9RLPN4fL6g2z/DFY3H4p4vDDvT7fX7DMwYoA=).

Reply:
> Symbolic links throw a FileNotFoundException.

Are you sure the link gets created on sharplab?

Reply:
Thanks for asking :)... That was my fault...

Do you know why the link does not exist? 
[new sharplab example](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmABgFgAoUgRmqLIAIjaA6ASQHkBua6gG4BDKEwBmASwA2MJgF4mABSEYAFmwDiMDABUYAWwAOAMWkwAckP0wAFAEo+VYaKkS8Aa3lKV6gMIR9YDdbZTVNbT0jUNV7NG8wrQwAJSE8ABMA0xlLa3sHfioJMRsAQiyYNgBRBAkAZwxam0kZOzsmAG9qJm6mcrZfWBVbZph8qh6WWgBOGwAiAZgh8TNZsYBfAqLSvuq6hptXD1aOrp7nJkP3DjwxCC8+haGAZQBPQIhXMAAZN3cD37iIzGE1YMwAJPNBhhZJcQB1LtdblUavVamtVo4NlQGNMbBCfh4mGAoTA0nD2jsUftLnZ0WNqH0ntovkJ6gB1KASaE6CS5S5xAAiQx51jY5ggAHc2ABBNJpACaiygjTgtFajhxMz6iRZ7M53N5thpY1BNm1zNZGA5XJgIuGZnVQA===)

Reply:
> Do you know why the link does not exist?
new sharplab example

@carlossanlop, or @Jozkee, can you answer this?

Can you also help us determine what the APIs do on Windows when the path is a symbolic link? See https://github.com/dotnet/runtime/pull/60507#discussion_r770537586.

Reply:
> Do you know why the link does not exist?

Not sure why sharplab fails; dotnetfiddle works fine, though, see https://dotnetfiddle.net/vFMxxf.

> Can you also help us determine what the APIs do on Windows when the path is a symbolic link? See #60507 (comment).

Depending whether you specify `FILE_FLAG_OPEN_REPARSE_POINT` in `FileOptions` or not, it should Get/Set the time in the link or the target. (That's my guess, haven't tested it but that's what I would expect).

Reply:
> Not sure why sharplab fails; dotnetfiddle works fine, though, see https://dotnetfiddle.net/vFMxxf.

![LinqPad Screenshot](https://user-images.githubusercontent.com/10373452/146656627-256ee0bc-f69c-46d5-8dc6-0bc0217c812a.png)
Same behavior on Windows

> Depending whether you specify FILE_FLAG_OPEN_REPARSE_POINT in FileOptions or not, it should Get/Set the time in the link or the target. (That's my guess, haven't tested it but that's what I would expect).

Thank you for clarification 👍🏼